auth0/v3/authentication/token_verifier.py

```diff
@@ -0,0 +1,225 @@
+import json
+import time
+
+import jwt
+import requests
+
+from auth0.v3.exceptions import TokenValidationError
+
+
+class SignatureVerifier():
+    DISABLE_JWT_CHECKS = {
+        "verify_signature": True,
+        "verify_exp": False,
+        "verify_nbf": False,
+        "verify_iat": False,
+        "verify_aud": False,
+        "verify_iss": False,
+        "require_exp": False,
+        "require_iat": False,
+        "require_nbf": False,
+    }
+
+    def __init__(self, algorithm):
+        if not algorithm or type(algorithm) != str:
+            raise ValueError("algorithm must be specified.")
+        self._algorithm = algorithm
+
+    def _fetch_key(self, key_id=None):
+        raise NotImplementedError
+
+    def verify_signature(self, token):
+        try:
+            header = jwt.get_unverified_header(token)
+        except jwt.exceptions.DecodeError:
+            raise TokenValidationError("ID token could not be decoded.")
+
+        alg = header.get('alg', None)
+        if alg != self._algorithm:
+            raise TokenValidationError('Signature algorithm of "{}" is not supported. Expected the ID token to be signed with "{}"'.format(alg, self._algorithm))
+
+        kid = header.get('kid', None)
+        secret_or_certificate = self._fetch_key(key_id=kid)
+
+        try:
+            decoded = jwt.decode(jwt=token, key=secret_or_certificate, algorithms=[self._algorithm], options=self.DISABLE_JWT_CHECKS)
+        except jwt.exceptions.InvalidSignatureError:
+            raise TokenValidationError("Invalid token signature.")
+        return decoded
+
+
+class SymmetricSignatureVerifier(SignatureVerifier):
+
+    def __init__(self, shared_secret, algorithm="HS256"):
```

Comment:
I'm not sure if the algorithm should be exposed here or kept hardcoded in the internals. I don't see any use case for auth0. Applies to the `AsymmetricSignatureVerifier` as well